### DISCUSSION TITLE
[mini.icons] Icon customization for a filename pattern

### DISCUSSION THREAD

**Discussion by tradiff** [nvim-mini/mini.nvim#1358] [General]
I'm trying to set a distinct icon for Ruby spec files to visually differentiate them from other ruby files. These files can be identified by:
- Filenames ending in `_spec.rb` (eg: `user_controller_spec.rb`)
- Being located in the `./spec/**/*` directory

I've explored the following config options:

- `extension`: Not applicable since `_spec.rb` isn't a true extension, and I don't want to affect all `*.rb` files.
- `file`: Works for specific files but doesn't support patterns.
- `filetype`: I could add a custom vim filetype. It seems very far-reaching, as it would require handling implications with treesitter, lsp, etc.

I'm curious if anyone has any other "outside the box" thoughts for this problem.

**Comment by echasnovski** [comment]
I am afraid 'mini.icons' was (deliberately) not really designed to make this kind of adjustment straightforward. The only way to have icon be decided based on pattern matching is indeed by using `vim.filetype.add()`. This is usually fine when target files don't already have filetype associated with them, which is not the case here.

One way to make it less painful for overall setup is to use custom filetype which starts with `'ruby.'`. This way it will source 'ruby' filetype and syntax plugins (see `:h 'filetype'`). It seems to work for tree-sitter highlighting (tested with for Lua files), but LSP might need extra work.

Here is the setup for you to try:

```lua
vim.filetype.add({
  pattern = {
    ['.*/spec/.*%.rb'] = 'ruby.spec',
    ['.*_spec%.rb'] = 'ruby.spec',
  },
})
require('mini.icons').setup({
  filetype = {
    ['ruby.spec'] = { glyph = '󰴭', hl = 'MiniIconsAzure' },
  },
  -- Make sure that .rb extension is not used during resolution and
  -- filetype matching is used instead
  use_file_extension = function(ext, _) return ext:sub(-2) ~= 'rb' end,
})
```

So what is left is to adjust LSP (is there anything else?) to be autostarted in 'ruby.spec' filetype also. But if you use `vim.lsp.start()` in 'after/filetype/ruby', then it might not need even that.

**Reply by tradiff** [reply]
Thanks @echasnovski (love your work btw, and Slava Ukraini!).

You're right, this is working, and was less painful than I expected. For anyone else trying to do something similar, I needed to add the new `ruby.spec` filetype to the lsp config:
```lua
lspconfig.solargraph.setup({
  on_attach = on_attach,
  capabilities = capabilities,
  filetypes = { "ruby", "ruby.spec" },
})
```

**Reply by tradiff** [reply]
This is not a problem for me, but just leaving some info for future explorers.

I show the directory matching mentioned above (`'.*/spec/.*%.rb'`) does not have any effect on mini.icons (icons in places like telescope, neotree, DiffView, and lualine). It does affect `:echo &filetype`, but mini.icons sees these files as `ruby`. Without looking at the source, I suspect mini.icons only cares about the filename, not the full path.

Again, not an issue for me, as the files I'm interested in do mostly match the `'.*_spec%.rb'` filename pattern, which works fine.

**Reply by echasnovski** [reply]
> I show the directory matching mentioned above (`'.*/spec/.*%.rb'`) does not have any effect on mini.icons (icons in places like telescope, neotree, DiffView, and lualine). It does affect `:echo &filetype`, but mini.icons sees these files as `ruby`. Without looking at the source, I suspect mini.icons only cares about the filename, not the full path.

That's exactly the opposite: 'mini.icons' does care about full path and [it is recommended](https://github.com/echasnovski/mini.nvim/blob/ccfc6300a5c17387f2899f979ecc04f942bee2a0/doc/mini-icons.txt#L43-L45) for plugins to supply full path to `MiniIcons.get()` (specifically for the reason you mention). The problem here is that those plugins don't have direct 'mini.icons' support in favor of 'nvim-tree/nvim-web-devicons'. If they use icons from 'mini.icons' for you, then it is probably thanks to calling [`mock_nvim_web_devicons()`](https://github.com/echasnovski/mini.nvim/blob/ccfc6300a5c17387f2899f979ecc04f942bee2a0/doc/mini-icons.txt#L461) in your config.

You can verify that 'mini.icons' respect the `'.*/spec/.*%.rb'` filetype matching pattern by something like `:=MiniIcons.get('file', '/home/user/spec/subdir/file.rb')` which should show you icon data configured for `'ruby.spec'` filetype.

By the way, all similar 'mini.nvim' modules ('mini.files', 'mini.pick', 'mini.statusline') work as expected here.

**Reply by tradiff** [reply]
Ah that makes sense. Thanks for adding that additional detail. So maybe those plugins will improve with time. As for me, I've been gradually swapping things over to mini where I can. I tend to customize pretty heavily, so often times mini is a bit too minimalist for my preferences 😅
